File: Java/HelloMicrosoft/src/main/java/com/hellomicrosoft/dao/CosmosClient.java
package com.hellomicrosoft.dao;

import com.hellomicrosoft.interfaces.ICosmosClient;
import com.hellomicrosoft.models.CounterDocument;
import org.springframework.stereotype.Component;

import java.util.concurrent.ConcurrentHashMap;

@Component
public class CosmosClient implements ICosmosClient {

    private final ConcurrentHashMap<String, CounterDocument> store = new ConcurrentHashMap<>();

    @Override
    public CounterDocument read(String id, String partitionKey) {
        int rand = (int)(Math.random() * 100) + 1;
        try { Thread.sleep(rand); } catch (InterruptedException e) { Thread.currentThread().interrupt(); } // simulate network latency
        String key = partitionKey + ":" + id;
        return store.getOrDefault(key, null);
    }

    @Override
    public void upsert(CounterDocument document, String matchETag) {
        int rand = (int)(Math.random() * 100) + 1;
        try { Thread.sleep(rand); } catch (InterruptedException e) { Thread.currentThread().interrupt(); } // simulate network latency
        String key = document.getPartitionKey() + ":" + document.getId();
        store.put(key, document);
    }
}

File: DotNet/HelloMicrosoft/Dao/CosmosClient.cs

public class CosmosClient : ICosmosClient
{
    private readonly Dictionary<string, CounterDocument> _store = new();

    public async Task<CounterDocument?> ReadAsync(string id, string partitionKey)
    {
        int rand = new Random().Next(1, 100);
        await Task.Delay(rand);  // simulate network latency
        var key = $"{partitionKey}:{id}";
        _store.TryGetValue(key, out var doc);
        return doc;
    }

    public async Task UpsertAsync(CounterDocument document, string? matchETag)
    {
        int rand = new Random().Next(1, 100);
        await Task.Delay(rand);  // simulate network latency
        var key = $"{document.partitionKey}:{document.id}";
        _store[key] = document;
        return;
    }
}
